### PR TITLE
Reserve excess capacity in seed cluster

### LINKED ISSUE #254: Reserve Excess Capacity in Seed Cluster
# Story
* As user I want my shoot cluster to be provisioned as fast as possible, so that I can work with the cluster right away (extremely useful for programmatic access and provisioning or for testing, training and demonstration purposes).

# Motivation
Initially we wanted to implement cluster excess capacity reservation in the mcm (gardener/machine-controller-manager#26), so that our seed clusters had sufficient capacity available for new shoot clusters. However, it seems easier to use low-priority pods to keep excess capacity available, which is also the general recommendation issued by the cluster autoscaler community that we are using. 

# Acceptance Criteria
- [ ] 3 new clusters or 5 % of the current number of clusters (whatever is larger) can be scheduled without provisioning of new machines in a seed cluster.

# Implementation Proposal
Let's use low priority pods (in exactly the configuration required for new shoot clusters) scheduled on the seed cluster. When a real shoot cluster control plane is scheduled with normal priorities, it will displace the low priority pods, but they will remain in the ledger and the cluster autoscaler will request new machines to fulfil the demand again. This will require the PriorityAdmissionPlugin. Since we most likely don't want to enable general configuration of admission plugins (preferred and most flexible solution, but may come with risks, that we would need to think through first), we can at least enable the PriorityAdmissionPlugin Gardener-wide (like the others we have enabled statically, which seems to have a limited impact/risk and makes this available for us in seed clusters, but also to our end users in shoot clusters).

For the entire discussion, take a look at gardener/machine-controller-manager#26.

# Definition of Done
- [ ] Knowledge is distributed: Have you spread your knowledge in pair programming/code review?
- [ ] Unit Tests are provided: Have you written automated unit tests or added manual NGPTT tickets?
- [ ] Integration Tests are provided: Have you written automated integration tests?
- [ ] Minimum API exposure: If you have added public API, was it really necessary/is it minimal?
- [ ] Operations guide: Have you updated the [operations guide](https://github.wdf.sap.corp/kubernetes/kube-docs/wiki/Operations%20Guide)?
- [ ] User documentation: Have you informed Sylvia Pick about user relevant changes?

### PR DESCRIPTION
**Pull request by prashanth26**
- This PR allows reservation of excess capacity to deploy new shoot control planes on a seed cluster
- Currently the shoot control plane size has been statically set to a requirement of 7CPU core and 14Gi of memory
- The number of shoot control planes that can be supported is 3 or 5% of currently running shoots, whichever is greater

Resolves: https://github.com/gardener/gardener/issues/254

TBD: Enable cluster-autoscaler on seed clusters to enable the working on reserve excess capacity

### CODE REVIEW

**Comment by prashanth26** [comment]
Hello @mvladev,
           Kindly have a look at this PR when you have time. This PR attempts to reserve excess capacity in the seed by adding a chart in the seed-bootstrap. 

Thanks & Regards,
Prashanth

**Comment by prashanth26** [comment]
Also, since we use priority classes which is beta in 1.11. Should we enable only for seeds with versions above 1.11 or even enable it for shoots below 1.11 (as alpha) ?

**Comment by vlerenc** [comment]
Our seeds are 1.10 right now and I don't think we should go to 1.11 already (too early, not enough experience yet). If it also works with 1.10, why not?

**Comment by prashanth26** [comment]
I have currently enabled it for all supported versions of kubernetes that is 1.8 and above. But the feature is only beta in 1.11, so I was just wondering if it's okay to take support from an alpha version feature to support gardener. It does seem to work properly in 1.10 as an alpha feature, I have tried it out.

**Comment by vlerenc** [comment]
Well, if you have tested it and no other workload was impacted, e.g. terminated (are all other pods in all namespaces running?), then I would assume it's good, alpha or not.

Please note the Kubernetes versions we currently run in the seeds:
- v1.10.5 for AWS, GCP, OpenStack
- v1.9.8 for Azure ‼️

**Comment by prashanth26** [comment]
I just tried with version 1.9.8 for Azure and seems to be working well. And yesterday I have already tried for 1.10.5 and 1.11.0. So i think it must be safe to say that it works properly across all k8s versions 1.9 and above.

**Comment by dkistner** [comment]
Have you calculated the amount of requested resources, which are required for all control plane pods? For each control plane would be 7 replicas with each 1000m cpu and 2Gi memory allocated, means in sum 7000m cpu and 14Gi memory per control plane. This is plenty.

I think also when you calculate with the resource limits, which are probably higher then the requests, these values would be pretty much. Ok we have to consider that we need some additional resources to schedule Terraform pods.

**Comment by prashanth26** [comment]
Thanks for your input @dkistner . Taking your observations into account, I have redone the following. I hope this looks better.

- I have summed the resource (requests/limits) for all control plane components (including one terraform pod running). 
- It turned out to be _cpu: 1765m/6600m, memory: 4166Mi/13350Mi_.

Hence I am updating the requests/limits to the following,
- Each deployment replica currently corresponds to resources of (request/limits) 300m/1000m of CPU and 700Mi/2048Mi of Memory. 
- ReplicasRequiredToSupportSingleShoot is 7.
- This corresponds to 2100m/7000m of CPU and 4900Mi/14Gi of RAM. 

Refer: https://github.com/gardener/gardener/pull/267/commits/3abbd365a22d9741a4422d8ba7ed0829b05e8b0b

**Comment by dkistner** [comment]
I think the limits will never reached by the pause containers, therefore I would only rely on the request, because they express how many resources are required to run a control plane with minimum requirements. I would suggest to allocate request/limits based on formula like this:
```
requests = (<sum of control-plane pods request> + <sum of resource requests for Terraform pods>) * 1.1
```


In your pause-container deployment you have currently set the requests to the same value as the limits.

**Comment by prashanth26** [comment]
I have updated the requests to the summation of cpu and memory requests. Refer previous comment and this commit - https://github.com/gardener/gardener/pull/267/commits/3abbd365a22d9741a4422d8ba7ed0829b05e8b0b. However the excess is not multiplied by 1.1, but is just a buffer value over summation request.

Requests Summation:
Memory: 4166Mi
CPU: 1765m

Total requests for Reserve Excess Capacity Deployment (with 7 replicas):
Memory: 4900Mi
CPU: 2100Mi

**Comment by prashanth26** [comment]
- Maximum resource usage _observed_ for a single pod in control plane is - CPU: ~200m, Memory: ~1150Mi
- Maximum resource requests for a single pod in control plane is - CPU: 200m, Memory: 1200Mi
- Total resource usage _observed_ for all pods in control plane is - CPU: ~500m, Memory: ~3000Mi
- Total resource requests for all pods in control plane is - CPU: 1765m, Memory: 4166Mi

The total resource requests for all pods is much greater than the actual usage in current control planes, and hence I am considering the total resource requests for the calculations that is - CPU: 1765m, Memory: 4166Mi (1), as the required resources for running all control plane pods.

- Keeping the above in mind, I am changing the reserve-excess-capacity deployment to have the following resource requirements per pod - CPU: 500m, Memory: 1200Mi  ----> so as to fit the largest pod by replacing the reserve-excess-capacity pod
- There would be 4 replicas of this pod to support each shoot cluster totalling to - CPU: 2000m, Memory: 4800Mi which is > (1) ----> so that all control plane pods fit comfortably 

Hope this sounds reasonable.

Refer: https://github.com/gardener/gardener/pull/267/commits/8fb3fafd99f9f45cfac0be5a047280c8f34f8d57

**Comment by dkistner** [comment]
The provided values matching with the values, which I have calculated: ``1665m`` cpu and ``3966Mi`` memory per control plane plus additional resources for the Terraform pods. 

Therefore I would agree having 4 replicas with each ``500m`` cpu + ``1200Mi`` memory seems to be sufficient.

Rest looks overall good to me. 
Would nevertheless waiting for the feedback of @mvladev

**Comment by prashanth26** [comment]
@dkistner Thanks for the feedback.

**Comment by hardikdr** [comment]
@prashanth26 can you please resolve the conflicting files?

Is there anything else to be discussed yet before merging?

**Comment by prashanth26** [comment]
@hardikdr I rebased the changes with the current master head. 

Thanks & Regards,
Prashanth

**Review by dkistner — APPROVED** [review]
As already mentioned, it looks good to me.
Asked @mvladev to have an additional look.

File: pkg/operation/hybridbotanist/controlplane.go
Comment (mvladev) [inline pkg/operation/hybridbotanist/controlplane.go]: Put those checks in

https://github.com/gardener/gardener/blob/92ca7287cf2cb7287410c6130fae4b32c9201dd4/charts/seed-controlplane/charts/kube-apiserver/templates/_helpers.tpl#L1-L5

File: pkg/operation/hybridbotanist/controlplane.go
Comment (mvladev) [inline pkg/operation/hybridbotanist/controlplane.go]: Put those checks in

https://github.com/gardener/gardener/blob/92ca7287cf2cb7287410c6130fae4b32c9201dd4/charts/seed-controlplane/charts/kube-scheduler/templates/_helpers.tpl#L1-L5

File: charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet-config-file
Comment (mvladev) [inline charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet-config-file]: Add this logic also in 
https://github.com/gardener/gardener/blob/92ca7287cf2cb7287410c6130fae4b32c9201dd4/charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet.flags#L25-L27

**Review by mvladev — CHANGES_REQUESTED** [review]

**Comment by prashanth26** [comment]
Hi @mvladev ,
                        I have made your suggested changes. All suggested changes are at - https://github.com/gardener/gardener/pull/267/commits/062e57bda20c41dce228a5c4159f28b5eb766de3.

Thanks & Regards,
Prashanth

File: pkg/operation/hybridbotanist/controlplane.go
Comment (prashanth26) [reply pkg/operation/hybridbotanist/controlplane.go]: Done. Refer - https://github.com/gardener/gardener/pull/267/commits/062e57bda20c41dce228a5c4159f28b5eb766de3?utf8=%E2%9C%93&diff=unified#diff-42520e4f41cf1180be16b2734375d5c5R5

File: pkg/operation/hybridbotanist/controlplane.go
Comment (prashanth26) [reply pkg/operation/hybridbotanist/controlplane.go]: Done. Refer - https://github.com/gardener/gardener/pull/267/commits/062e57bda20c41dce228a5c4159f28b5eb766de3?utf8=%E2%9C%93&diff=unified#diff-a422a9d10a682f8e4a4c20051076b849R5

File: charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet-config-file
Comment (prashanth26) [reply charts/shoot-cloud-config/charts/original/templates/kubelet/_kubelet-config-file]: Done. Refer - https://github.com/gardener/gardener/pull/267/commits/062e57bda20c41dce228a5c4159f28b5eb766de3?utf8=%E2%9C%93&diff=unified#diff-7a590da443a0105c0f781f2e0fd2c726R28

**Review by mvladev — APPROVED** [review]
lgtm

**Comment by prashanth26** [comment]
Thank you.

**Pull request merged**
